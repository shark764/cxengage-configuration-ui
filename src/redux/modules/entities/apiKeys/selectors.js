import { Map } from 'immutable';
import { getSelectedEntity } from '../selectors';
import { selectTenantRoles } from '../roles/selectors';

export const getInitialUpdateFormValues = state =>
  new Map({
    name: getSelectedEntity(state).get('name'),
    description: getSelectedEntity(state).get('description'),
    roleId: getSelectedEntity(state).get('roleId'),
    id: getSelectedEntity(state).get('id'),
    secret: getSelectedEntity(state).get('secret')
  });

export const getInitialCreateFormValues = state =>
  new Map({
    roleId: selectTenantRoles(state).length > 0 ? selectTenantRoles(state)[0].value : ''
  });
