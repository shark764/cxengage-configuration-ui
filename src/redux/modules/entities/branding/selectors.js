/*
 * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
 */

export const getBranding = state =>
  state.getIn(['Entities', 'branding', 'data']);

export const getProtectedBranding = state =>
  state.getIn(['Entities', 'protectedBranding', 'data']);
