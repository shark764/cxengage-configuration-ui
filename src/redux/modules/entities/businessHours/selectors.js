import { selectFormInitialValues } from '../../form/selectors';
import { getSelectedEntity } from '../selectors';
import { Map } from 'immutable';

export const selectBusinessHoursFormInitialValues = state => {
  if (getSelectedEntity(state) === undefined) {
    const initialTimes = ['mon', 'tue', 'wed', 'thu', 'fri', 'sat', 'sun'].reduce(
      (r, key) => ({ ...r, [key.concat('StartTimeMinutes')]: 0, [key.concat('EndTimeMinutes')]: 0 }),
      {}
    );
    return new Map({ active: true, businessHoursType: '24/7', timezone: 'Etc/GMT+12', ...initialTimes });
  }
  return selectFormInitialValues(state);
};
