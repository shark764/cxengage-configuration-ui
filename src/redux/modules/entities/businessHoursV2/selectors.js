import { createSelector } from 'reselect';
import { Map, List } from 'immutable';

import { selectFormInitialValues } from '../../form/selectors';
import { getSelectedEntity } from '../selectors';

export const selectBusinessHoursV2FormInitialValues = state => {
  if (getSelectedEntity(state) === undefined) {
    return Map({ active: true, shared: false });
  }
  return selectFormInitialValues(state);
};

export const getBusinessHourV2Drafts = createSelector(
  getSelectedEntity,
  businessHour =>
    businessHour &&
    businessHour.get('drafts') &&
    List.isList(businessHour.get('drafts')) &&
    businessHour.get('drafts').toJS()
);
