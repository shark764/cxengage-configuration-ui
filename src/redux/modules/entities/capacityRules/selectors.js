import { createSelector } from 'reselect';
import { Map, List, fromJS } from 'immutable';

import { selectFormInitialValues } from '../../form/selectors';
import { getSelectedEntity, getSelectedSubEntity } from '../selectors';
import { ednRuleToImmutable } from './ednHelpers';

export const selectCapacityRulesFormInitialValues = (state) => {
  if (getSelectedEntity(state) === undefined) {
    return Map({ name: '' });
  }
  return selectFormInitialValues(state);
};

export const getCapacityRules = (state) => state.getIn(['Entities', 'capacityRules', 'data']);

export const getCapacityRulesSelector = createSelector([getCapacityRules], (capacityRules) =>
  capacityRules.filter((rule) => rule.get('active')).map((rule) => ({ label: rule.get('name'), value: rule.get('id') }))
);

export const selectCapacityRuleVersions = createSelector(
  getSelectedEntity,
  (capacityRule) =>
    capacityRule
      ? capacityRule
          .get('items', List([]))
          .map((capacityRule, i) => capacityRule.set('numericOrderVersion', `v${i + 1}`))
          .toJS()
      : List([]).toJS()
);

export const selectCapacityRuleVersionFormInitialValues = createSelector(
  getSelectedSubEntity,
  (version) =>
    version
      ? version.get('ruleSet') && version.set('rule', ednRuleToImmutable(version.get('ruleSet')))
      : fromJS({
          rule: {
            voice: true,
            groups: [
              {
                channels: ['email', 'messaging'],
                interactions: 4,
              },
            ],
          },
          quantifier: 'any',
        })
);
