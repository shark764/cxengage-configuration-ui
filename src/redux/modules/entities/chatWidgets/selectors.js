/*
 * Copyright Â© 2015-2020 Serenova, LLC. All rights reserved.
 */

import { Map, List } from 'immutable';
import { createSelector } from 'reselect';
import { getSelectedEntity } from '../selectors';
import { selectFormInitialValues, getCurrentFormValueByFieldName } from '../../form/selectors';

export const selectChatWidgetFormInitialValues = state => {
  if (getSelectedEntity(state) === undefined) {
    return new Map({
      brandColor: '#65758e',
      conversationColor: '#0099ff',
      actionColor: '#0099ff',
      displayStyle: 'button',
      buttonWidth: '58',
      buttonHeight: '58',
      prechatCapture: 'name'
    });
  }
  return selectFormInitialValues(state);
};

const getChatWidgets = state => state.getIn(['Entities', 'chatWidgets', 'data'], new List([]));
const getDigitalChannelsApps = state => state.getIn(['Entities', 'digitalChannelsApps', 'data'], new List([]));

export const getDigitalChannelsAppIds = createSelector(
  [getDigitalChannelsApps, getChatWidgets],
  (digitalChannelsApps, chatWidgets) =>
    digitalChannelsApps
      .filter(app => chatWidgets.filter(cw => cw.get('appId') === app.get('id')).size === 0)
      .map(app => ({ value: app.get('id'), label: app.get('name') }))
);

export const getDigitalChannelsApp = state => {
  return (
    getSelectedEntity(state) &&
    getSelectedEntity(state).get('appId') &&
    getDigitalChannelsApps(state).find(app => app.get('id') === getSelectedEntity(state).get('appId'))
  );
};

export const getDisplayStyleIsButton = state => getCurrentFormValueByFieldName(state, 'displayStyle') === 'button';
