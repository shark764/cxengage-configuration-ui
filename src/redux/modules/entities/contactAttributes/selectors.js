/*
 * Copyright Â© 2015-2020 Serenova, LLC. All rights reserved.
 */

import { createSelector } from 'reselect';
import { getEntities } from '../selectors';

export const getContactAttributes = createSelector(
  getEntities,
  entitites => entitites && entitites.getIn(['contactAttributes', 'data'])
);

export const getContactAttributesNames = createSelector(
  [getContactAttributes],
  contactAttributes =>
    contactAttributes && contactAttributes.size > 0
      ? contactAttributes
          .toJS()
          .map(attribute => attribute.objectName)
          .sort((a, b) => (a.toUpperCase() < b.toUpperCase() ? -1 : 0))
      : undefined
);

export const getMandatoryContactAttributes = createSelector(
  [getContactAttributes],
  contactAttributes => contactAttributes && contactAttributes.filter(attribute => attribute.get('mandatory') === true)
);
