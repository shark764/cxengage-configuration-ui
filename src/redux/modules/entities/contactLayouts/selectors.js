/*
 * Copyright Â© 2015-2020 Serenova, LLC. All rights reserved.
 */

import { Map, List, fromJS } from 'immutable';
import { createSelector } from 'reselect';
import { generateUUID } from 'serenova-js-utils/uuid';

import { getCurrentFormValues, isFormDirty } from '../../form/selectors';
import { getEntities, getSelectedEntity, getSelectedSubEntityId, getSelectedEntityId } from '../selectors';
import {
  getContactAttributes,
  getContactAttributesNames,
  getMandatoryContactAttributes
} from '../contactAttributes/selectors';

export const getContactLayoutsFormInitialValues = createSelector(
  [getSelectedEntity, getSelectedEntityId],
  (selectedEntity, selectedEntityId) =>
    selectedEntityId === 'create'
      ? Map({})
      : Map({
          name: selectedEntity.get('name'),
          description: selectedEntity.get('description'),
          active: selectedEntity.get('active')
        })
);

// Arrange contact-layout items that belong to the same category under the first category-item in each list:
export const selectContactLayoutsHeaders = createSelector([getCurrentFormValues], values => {
  if (values && values.get('layout') && values.get('layout').size > 0) {
    return values.get('layout').reduce((accumlator, currentLayout) => {
      const hierarchyExists = accumlator.find(val => val.get('hierarchy') === currentLayout.get('hierarchy'));
      if (!hierarchyExists) {
        return accumlator.push(currentLayout);
      }
      return accumlator;
    }, List());
  } else {
    return List();
  }
});

export const isCurrentFormMissingMandatoryAttributes = createSelector(
  [getCurrentFormValues, getMandatoryContactAttributes],
  (currentFormValues, mandatoryAttributes) =>
    currentFormValues && currentFormValues.get('layout')
      ? mandatoryAttributes &&
        mandatoryAttributes.some(
          mandatoryAttribute =>
            currentFormValues
              .get('layout')
              .find(layout => layout.get('contactAttributeId') === mandatoryAttribute.get('id')) === undefined
        )
      : false
);

export const getMissingMandatoryAttributesNames = createSelector(
  [getCurrentFormValues, getMandatoryContactAttributes],
  (currentFormValues, mandatoryAttributes) => {
    if (currentFormValues && currentFormValues.get('layout')) {
      const missingMandatoryAttributes =
        mandatoryAttributes &&
        mandatoryAttributes.filter(
          mandatoryAttribute =>
            currentFormValues
              .get('layout')
              .find(layout => layout.get('contactAttributeId') === mandatoryAttribute.get('id')) === undefined
        );
      return missingMandatoryAttributes ? missingMandatoryAttributes.map(attr => attr.get('objectName')) : List();
    } else {
      return List();
    }
  }
);

export const getExistingCategories = createSelector(
  [getCurrentFormValues],
  currentFormValues =>
    currentFormValues && currentFormValues.get('layout')
      ? currentFormValues.get('layout').reduce((accumlator, currentLayout) => {
          const hierarchyIndex = accumlator.findIndex(
            val => val.get('categoryUUID') === currentLayout.get('categoryUUID')
          );
          return hierarchyIndex === -1
            ? accumlator.push(
                fromJS({ hierarchy: currentLayout.get('hierarchy'), categoryUUID: currentLayout.get('categoryUUID') })
              )
            : accumlator;
        }, List())
      : List()
);

export const getExistingCategoryNamesInCurrentLayout = createSelector(
  [getExistingCategories],
  existingCategories =>
    existingCategories && existingCategories.size > 0 ? existingCategories.map(a => a.get('hierarchy')) : List()
);

export const getContactLayoutsFormSubmitValues = createSelector(
  [getCurrentFormValues, getSelectedEntityId],
  (currentFormValues, selectedEntityId) => {
    const updatedValues = currentFormValues.update('layout', layout =>
      layout.reduce((accumlator, currentLayout) => {
        const hierarchyIndex = accumlator.findIndex(
          val => (val.getIn(['label', 'name']) || val.getIn(['label', 'en-US'])) === currentLayout.get('hierarchy')
        );
        if (hierarchyIndex === -1) {
          return accumlator.push(
            fromJS({
              label: currentLayout.get('label'),
              attributes: [currentLayout.get('contactAttributeId')]
            })
          );
        } else {
          return accumlator.updateIn([hierarchyIndex, 'attributes'], attributes =>
            attributes.push(currentLayout.get('contactAttributeId'))
          );
        }
      }, List())
    );
    return fromJS({
      name: updatedValues.get('name'),
      description: updatedValues.get('description') ? updatedValues.get('description') : '',
      active: selectedEntityId === 'create' ? false : updatedValues.get('active'),
      layout: updatedValues.get('layout')
    });
  }
);

// Contact-Attributes must be unique in a layout, filter out the attributes that were already selected:
export const getAvailableContactAttributesNames = createSelector(
  [getContactAttributesNames, getCurrentFormValues],
  (attributeNames, formValues) =>
    attributeNames &&
    attributeNames.filter(
      attrName => formValues.get('layout').find(layout => layout.get('name') === attrName) === undefined
    )
);

export const getContactLayoutsSubEntityFormInitialValues = createSelector(
  [getCurrentFormValues, getSelectedSubEntityId],
  (currentFormValues, selectedSubEntityId) => {
    if (selectedSubEntityId === 'create') {
      return fromJS({
        label: [{ label: '', language: '' }],
        categoryUUID: generateUUID(),
        droppableUUID: generateUUID(),
        draggableUUID: generateUUID(),
        endpointUUID: generateUUID()
      });
    } else {
      const categoryUUID = selectedSubEntityId.startsWith('createListItem')
        ? selectedSubEntityId.replace('createListItem:', '')
        : selectedSubEntityId.replace('updateCategoryHeader:', '');

      // select the first category item in the list:
      const layout =
        currentFormValues &&
        currentFormValues.get('layout') &&
        currentFormValues.get('layout').find(layout => layout.get('categoryUUID') === categoryUUID);

      // convert label object to array & remove label item that have key "name":
      const values = layout
        ? layout
            .update('label', label => label.reduce((r, v, k) => r.push(fromJS({ label: v, language: k })), List()))
            .update('label', label => label.filter(a => a.get('language') !== 'name'))
        : List();

      // set draggableUUID, endpointUUID & remove name key from the values while creating a new list-item:
      return selectedSubEntityId.startsWith('createListItem:')
        ? values
            .set('draggableUUID', generateUUID())
            .set('endpointUUID', generateUUID())
            .delete('name')
        : values;
    }
  }
);

export const getContactLayoutsSubEntityFormSubmitValues = (state, props) => {
  if (props.values && props.values.size > 0) {
    // change the label array back to object & update the name key for the category localization:
    const updatedValues = props.values
      .update('label', label =>
        label.reduce(
          (r, v) => (v.get('language') !== '' && v.get('label') !== '' ? r.set(v.get('language'), v.get('label')) : r),
          Map({})
        )
      )
      .update('label', label => label.set('name', props.values.get('hierarchy')));

    // update the contactAttributeId to match the selected contactAttribute:
    const attribute =
      getContactAttributes(state) &&
      getContactAttributes(state).find(attr => attr.get('objectName') === updatedValues.get('name'));

    const subFormValues = attribute ? updatedValues.set('contactAttributeId', attribute.get('id')) : updatedValues;

    return getCurrentFormValues(state).update('layout', layout => {
      if (props.selectedSubEntityId.startsWith('create')) {
        return layout.push(subFormValues);
      } else {
        return layout.map(
          a =>
            a.get('categoryUUID') === subFormValues.get('categoryUUID')
              ? a.set('label', subFormValues.get('label')).set('hierarchy', subFormValues.get('hierarchy'))
              : a
        );
      }
    });
  } else {
    return Map({});
  }
};

export const getActiveContactLayouts = createSelector(
  getEntities,
  entitites => (entitites ? entitites.getIn(['contactLayouts', 'data']).filter(a => a.get('active') === true) : [])
);

export const shouldDisableContactLayoutsStatusToggle = createSelector(
  [isCurrentFormMissingMandatoryAttributes, getSelectedEntity, getActiveContactLayouts, isFormDirty],
  (missingMandatoryAttributes, selectedEntity, activeContactLayouts, isFormDirty) =>
    isFormDirty ||
    missingMandatoryAttributes ||
    (selectedEntity && selectedEntity.get('active') === true ? activeContactLayouts.size === 1 : false)
);
