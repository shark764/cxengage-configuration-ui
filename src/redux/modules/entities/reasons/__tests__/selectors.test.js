/*
 * Copyright Â© 2015-2018 Serenova, LLC. All rights reserved.
 */

import { fromJS } from 'immutable';
import { selectReasonsFormInitialValues } from '../selectors';
import { getSelectedEntity, getSelectedEntityId } from '../../selectors';
import { selectFormInitialValues } from '../../../form/selectors';

jest.mock('../../../form/selectors');
jest.mock('../../selectors');

describe('selectReasonsFormInitialValues', () => {
  it('Return the initial Reason Values, all values provided', () => {
    const initialState = fromJS({
      id: 'mockId',
      shared: false,
      active: true
    });
    selectFormInitialValues.mockImplementation(() => initialState);
    getSelectedEntity.mockImplementation(() => initialState);
    getSelectedEntityId.mockImplementation(() => 'mockId');
    expect(selectReasonsFormInitialValues()).toMatchSnapshot();
  });

  it('Empty Reason, all values provided', () => {
    const initialState = fromJS({
      id: 'mockId',
      shared: false,
      active: true
    });
    selectFormInitialValues.mockImplementation(() => initialState);
    getSelectedEntity.mockImplementation(() => initialState);
    getSelectedEntityId.mockImplementation(() => 'mockId');
    expect(selectReasonsFormInitialValues()).toMatchSnapshot();
  });

  it('Return the initial Reason Values, all values NOT provided', () => {
    const initialState = fromJS({
      id: 'mockId'
    });
    selectFormInitialValues.mockImplementation(() => initialState);
    expect(selectReasonsFormInitialValues()).toMatchSnapshot();
  });
});
