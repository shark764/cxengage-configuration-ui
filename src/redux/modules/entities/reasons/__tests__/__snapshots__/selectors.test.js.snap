// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectReasonsFormInitialValues Empty Reason, all values provided 1`] = `
Immutable.Map {
  "id": "mockId",
  "shared": false,
  "active": true,
}
`;

exports[`selectReasonsFormInitialValues Return the initial Reason Values, all values NOT provided 1`] = `
Immutable.Map {
  "id": "mockId",
}
`;

exports[`selectReasonsFormInitialValues Return the initial Reason Values, all values provided 1`] = `
Immutable.Map {
  "id": "mockId",
  "shared": false,
  "active": true,
}
`;
